### PR TITLE
Add shaking to randomize

### LINKED ISSUE #188: Shake to randomize
As a Buyer, I want to shake my phone to randomize the random item page, because I like having fun

### PR DESCRIPTION
**Pull request by cronokirby**
Fixes #188

### CODE REVIEW

File: app/src/main/java/com/example/sharingang/shake/ShakeDetector.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    fun detect(x: Float, y: Float, z: Float, now: Long): Boolean {
+        val gForce = sqrt(x * x + y * y + z * z) / SensorManager.GRAVITY_EARTH
+        // Ignore shakes without enough force
+        if (gForce < SHAKE_THRESHOLD_GRAVITY) {
+            return false
+        }
+        // Ignore shakes that happen too close to each other
+        if (shakeTimestamp + SHAKE_WAIT > now) {
+            return false
+        }
```
Comment (lumapools) [inline app/src/main/java/com/example/sharingang/shake/ShakeDetector.kt:45]: I think you could merge these two `if`s

File: app/src/test/java/com/example/sharingang/shake/ShakeDetectorTest.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.junit.Test
+
+class ShakeDetectorTest {
+    @Test
+    fun noShakeIsDetectedForSmallMovements() {
+        val detector = ShakeDetector()
+        assert(!detector.detect(0F, 0F, 0F, 0))
+        assert(!detector.detect(0F, 0F, 0F, 10000))
+    }
```
Comment (lumapools) [inline app/src/test/java/com/example/sharingang/shake/ShakeDetectorTest.kt:11]: Does this test case cover the case when there is no movement at all? I think you could maybe add a test case for when there is movement but not enough to trigger the shake condition.

File: app/src/main/java/com/example/sharingang/MainActivity.kt [hunk 3]
```diff
@@ -78,17 +106,23 @@ class MainActivity : AppCompatActivity(R.layout.activity_main) {
...
-                val random = allItems.random()
+
+                val random = itemRepository.getAll().random()
                 lifecycleScope.launch(Dispatchers.Main) {
                     val drawerLayout = findViewById<DrawerLayout>(R.id.drawer_layout)
                     drawerLayout.close()
                     getNavController().navigate(
                         ItemsListFragmentDirections
-                            .actionItemsListFragmentToDetailedItemFragment(random)
+                            .actionGlobalDetailedItemFragment(random)
```
Comment (lumapools) [inline app/src/main/java/com/example/sharingang/MainActivity.kt:125]: I am not sure I understand this change. What is the difference between `action...To...` and `actionGlobal...`?

File: app/src/main/java/com/example/sharingang/MainActivity.kt [hunk 4]
```diff
@@ -78,17 +106,23 @@ class MainActivity : AppCompatActivity(R.layout.activity_main) {
...
+        selectRandomItem()
+    }
+
+    private fun selectRandomItem() {
         lifecycleScope.launch(Dispatchers.IO) {
             val allItems = itemRepository.getAll()
             if (allItems.isNotEmpty()) {
-                val random = allItems.random()
+
+                val random = itemRepository.getAll().random()
```
Comment (lumapools) [inline app/src/main/java/com/example/sharingang/MainActivity.kt:119]: Why do we need to `getAll()` again instead of keeping `allItems.random()`?

**Review by lumapools — CHANGES_REQUESTED** [review]
Awesome, well done! 
There are just a few things that I did not fully understand but this definitely is a fun feature! ^^

Edit: Oops I just noticed this is a Draft, I don't know if it's fine to request changes in a WIP.

File: app/src/main/java/com/example/sharingang/MainActivity.kt (hunk 4 above)
Comment (v1ruso) [reply app/src/main/java/com/example/sharingang/MainActivity.kt:119]: Not sure I understand why either

File: app/src/main/java/com/example/sharingang/MainActivity.kt [hunk 5]
```diff
@@ -24,6 +28,10 @@ class MainActivity : AppCompatActivity(R.layout.activity_main) {
     @Inject
     lateinit var itemRepository: ItemRepository
 
+    lateinit var shakeListener: ShakeListener
+    lateinit var sensorManager: SensorManager
+    lateinit var accelerometer: Sensor
```
Comment (v1ruso) [inline app/src/main/java/com/example/sharingang/MainActivity.kt:33]: Do you need this to be public?

File: app/src/test/java/com/example/sharingang/shake/ShakeDetectorTest.kt (hunk 2 above)
Comment (cronokirby) [reply app/src/test/java/com/example/sharingang/shake/ShakeDetectorTest.kt:11]: Yeah, not a bad idea for small movement

File: app/src/main/java/com/example/sharingang/MainActivity.kt (hunk 3 above)
Comment (cronokirby) [reply app/src/main/java/com/example/sharingang/MainActivity.kt:125]: The former requires us to be on the items list fragment. This crashes if you're not there, e.g. shaking from somewhere else. Using a global action makes us navigate regardless of where we happen to be.

File: app/src/main/java/com/example/sharingang/MainActivity.kt (hunk 5 above)
Comment (cronokirby) [reply app/src/main/java/com/example/sharingang/MainActivity.kt:33]: woops

File: app/src/main/java/com/example/sharingang/shake/ShakeDetector.kt (hunk 1 above)
Comment (cronokirby) [reply app/src/main/java/com/example/sharingang/shake/ShakeDetector.kt:45]: Sure, but I feel like it's clearer if the logic is separate

File: app/src/main/java/com/example/sharingang/MainActivity.kt (hunk 4 above)
Comment (cronokirby) [reply app/src/main/java/com/example/sharingang/MainActivity.kt:119]: Mistake when merging :P

File: app/src/main/java/com/example/sharingang/MainActivity.kt (hunk 3 above)
Comment (lumapools) [reply app/src/main/java/com/example/sharingang/MainActivity.kt:125]: Okay, thanks!

**Review by lumapools — APPROVED** [review]
Nice!

**Review by v1ruso — APPROVED** [review]
Well done:)

**Pull request merged**